src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1625,34 +1629,66 @@ RouterJob(Query *originalQuery, PlannerRestrictionContext *plannerRestrictionCon
 
 	if (originalQuery->commandType == CMD_SELECT)
 	{
-		job->taskList = SingleShardSelectTaskList(originalQuery, relationShardList,
-												  placementList, shardId);
+		job->taskList = SingleShardSelectTaskList(originalQuery, job->jobId,
+												  relationShardList, placementList,
+												  shardId);
+
+		/*
+		 * Queries to reference tables, or distributed tables with multiple replica's have
+		 * their task placements reordered according to the configured
+		 * task_assignment_policy. This is only applicable to select queries as the modify
+		 * queries will be reordered to _always_ use the first-replica policy during
+		 * execution.
+		 */
+		ReorderTaskPlacementsByTaskAssignmentPolicy(job, TaskAssignmentPolicy);
 	}
 	else if (isMultiShardModifyQuery)
 	{
-		job->taskList = QueryPushdownSqlTaskList(originalQuery, 0,
+		job->taskList = QueryPushdownSqlTaskList(originalQuery, job->jobId,
 												 plannerRestrictionContext->
 												 relationRestrictionContext,
 												 relationShardList, MODIFY_TASK,
 												 requiresMasterEvaluation);
 	}
 	else
 	{
-		job->taskList = SingleShardModifyTaskList(originalQuery, relationShardList,
-												  placementList, shardId);
+		job->taskList = SingleShardModifyTaskList(originalQuery, job->jobId,
+												  relationShardList, placementList,
+												  shardId);
 	}
 
 	job->requiresMasterEvaluation = requiresMasterEvaluation;
 	return job;
 }
 
 
+/*
+ * ReorderTaskPlacementsByTaskAssignmentPolicy applies selective reordering for supported
+ * TaskAssignmentPolicyTypes.
+ *
+ * Supported Types
+ * - TASK_ASSIGNMENT_ROUND_ROBIN round robin schedule queries among placements
+ *
+ * By default it does not reorder the task list, implying a first-replica strategy.
+ */
+static void
+ReorderTaskPlacementsByTaskAssignmentPolicy(Job *job, TaskAssignmentPolicyType
```

Comment:
This function looked more like it belonged to multi_physical_planner with the rest of task assignment functions

Reply:
I thought about this as well, however this function is a very router planner specific version to `AssignAnchorShardTaskList `: https://github.com/citusdata/citus/blob/89bc3ad5cbd2c4616840439f6de6d244f31a5f0d/src/backend/distributed/planner/multi_physical_planner.c#L4903-L4924

Unfortunately there were all kinds of side effects when reusing that function when the policy was not set to `round-robin`. Given that that function was not reusable for this feature and this is very specific to the router planner I am inclined to keep it here.